aten/src/ATen/FunctionalStorageImpl.cpp

```diff
@@ -0,0 +1,110 @@
+#include <ATen/FunctionalStorageImpl.h>
+
+#include <ATen/FunctionalTensorWrapper.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+#include <c10/util/Exception.h>
+#include <vector>
+
+namespace at {
+namespace functionalization {
+
+Alias::Alias(const at::Tensor& base) {
+  TORCH_INTERNAL_ASSERT(!at::functionalization::impl::isFunctionalTensor(base));
+  base_ = base;
+}
+
+const at::Tensor& Alias::base() const {
+  return base_;
+}
+
+// metas is taken by value on purpose - we want to copy the vector.
+void Alias::add_update(const at::Tensor& updated_val, std::vector<ViewMeta> metas) {
+  updates_.push_back({updated_val, metas});
+  generation_++;
+}
+
+// Note [Functionalization: Alias Removal Part 2]
+// See Note [Functionalization: Alias Removal] for more details.
+// This function applies a single update from one of the views to the Alias object.
+// We start out with <original_base> and <mutated_view>, and our goal is to end up with <mutated_base>.
+void Alias::apply_update(const Update& update) {
+  at::AutoDispatchBelowFunctionalize guard;
+  at::Tensor t = update.new_val;
+  TORCH_INTERNAL_ASSERT(!at::functionalization::impl::isFunctionalTensor(t));
+  std::vector<at::Tensor> tmp_values({base_});
+  // First, we replay each view on <original_base>.
+  // We only actually need this for ops like select/slice/diagonal, which creates a view that's a subset of the original tensor.
+  // e.g.:
+  // a = torch.ones(...)
+  // b = a[0]
+  // c = a.permute(...)
+  // d = c[0]
+  // d.add_(1)
+  // In order to get the mutated version of a, we need to know the intermediates b and c.
+  for (size_t i = 0; i < update.view_metas.size() - 1; ++i) {
```

Comment:
Testing `i < update.view_metas.size() - 1` means you will skip the last view_meta. I think this is probably intentional but it is worth calling out.

Reply:
yep that's right. I'll update the comment